README
Make sure lm_sensors is installed, it will not work without it.

You might want to take a look at the Makefile to change an include path or
library path.

* Install with the following commands:

    make
    su
    make install


The default install path is /usr/local/bin, if you do not want this, use

    make DEST=/to/another/path install

to install wmtemp to /to/an/other/path.

To install the manual page to another location (default is
/usr/local/share/man/man1), use:

    make MAN=/to/another/path install

Or combine the above :).

* Check out `wmtemp --help` for options.

* See the file COPYING for license information


ChangeLog
0.0.6 (2008-04-28)
* Riccardo Stagni <unriccio@email.it> (the current Debian maintainer of this
  package) submitted a patch fixing some syntactic stuff in the -h output

0.0.5 (2005-06-23)
* Found out Debian (http://www.debian.org/) has a package and the maintainer
  has added a manual page, which I've shamelessly copied. Only the part
  explaining this package had no manual page is removed, since it's no longer
  true :). Thanks to Lars Steinke <lss@debian.org> for maintaining the Debian
  package.

0.0.4 (2004-03-12)
* Fixed a small bug in the Makefile, probably caused by an older version of
  sed, which caused gcc to complain about extra characters after an #include
  when compiling the xpms. Again thanks to Jean Delvare for pointing this out.

0.0.3 (2004-03-10)
* Modified the background_off.xpm image a bit, so it would be more consistent
  with WMCPULoad and WMMemLoad (more shadow at the borders)
* Added a flag to show temperatures in Fahrenheit
* Added a flag to show temperatures in Kelvin
* Modified the digit code so it could show values greater than 100
* Modified the build system, so XPMs are build seperately instead of included
  in the main.c file
* Stripped all long arguments, they are quite useless and take more effort to
  type
* Do not wait <interval> seconds before the first update, just update, then
  enter the main loop.
* Removed the optimization options for Pentium 4, not everybody has a Pentium
  4 to play with :). Also made de CFLAGS a bit more conservative.
* Some textual changes in the help output, thanks to Jean Delvare
  (http://www.ensicaen.ismra.fr/~delvare/) for pointing that out.

0.0.2 (2004-03-08)
* Added a couple of options, so CPU and SYS features could be swapped or
  entirely customized
* Added README, COPYING and AUTHORS files and made the package

0.0.1 (2004-03-08)
* Stripped WMMemLoad to the bones and replaced the memory statistics code
  with temperature statistics code, obtained from libsensors (a part of
  the lm_sensors package


temp.c
#define	DEBUG(x) x;
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <sensors.h>
#include <sensors/error.h>

#include "temp.h"

sensor_path_t cpu_sensor_path = { NULL, NULL, NULL };
sensor_path_t sys_sensor_path = { NULL, NULL, NULL };

temperature_t t_type = CELCIUS;

typedef struct sensor_source_t {
    int valid;
    sensors_chip_name const *chip;
    int feature;
} sensor_source_t;

static FILE *f = NULL;
static char sensors = 0;

static sensor_source_t cpu_feature = {0, NULL, 0};
static sensor_source_t sys_feature = {0, NULL, 0};

void temp_set_default(sensor_path_t *path, char const *subfeature)
{
    if (!(path->chip || path->feature || path->subfeature)) {
        path->subfeature = subfeature;
    }
}

void temp_deinit() {
	if (f != NULL) {
		fclose(f);
	}

	if (sensors) {
		sensors_cleanup();
	}
}

void temp_find_feature(sensor_path_t const *path, sensor_source_t *feature)
{
	const sensors_chip_name *s_name;
    const sensors_feature *s_feature;
	const sensors_subfeature *s_subfeature;
    char *s_feat_name;
	char str[256];
    int chip_nr = 0, f_nr;

	while ((s_name = sensors_get_detected_chips(NULL, &chip_nr)) != NULL) {
		sensors_snprintf_chip_name(str, 256, s_name);
        DEBUG(printf("L%d - chip_nr %d, name %s\n", __LINE__, chip_nr, str));
        if (path->chip != NULL && strcasecmp(str, path->chip)) {
            continue;
        }

        f_nr = 0;
        while ((s_feature = sensors_get_features(s_name, &f_nr)) != NULL) {
		    s_feat_name = sensors_get_label(s_name, s_feature);
            DEBUG(printf("L%d - feature #%d : %s\n", __LINE__, f_nr, s_feat_name));
            if (path->feature != NULL && strcasecmp(s_feat_name, path->feature)) {
                continue;
            }

            s_subfeature = sensors_get_subfeature(s_name, s_feature,
                            SENSORS_SUBFEATURE_TEMP_INPUT);
		    if (!s_subfeature) {
                continue;
            }
            DEBUG(printf("L%d - subfeature %s\n", __LINE__, s_subfeature->name));

            if (path->subfeature == NULL
                    || !strcasecmp(s_subfeature->name, path->subfeature)) {
                feature->valid = 1;
                feature->chip = s_name;
                feature->feature = s_subfeature->number;
                DEBUG(printf("L%d - found\n", __LINE__));
                return;
            }
        }
    }
}

void temp_init(const char *filename) {
    temp_set_default(&cpu_sensor_path, "temp1_input");
    temp_set_default(&sys_sensor_path, "temp2_input");

	atexit(temp_deinit);
	if (filename) {
		f = fopen(filename, "r");
		if (f == NULL) {
			fprintf(stderr, "could not open configfile %s: %s\n", filename,
					strerror(errno));
			exit(1);
		}
	}

	if (sensors_init(f)) {
		fprintf(stderr, "could not initialize sensors\n");
		exit(1);
	}

	sensors = 1;
    temp_find_feature(&cpu_sensor_path, &cpu_feature);
    temp_find_feature(&sys_sensor_path, &sys_feature);
}

unsigned int temp_read(sensor_source_t const *source) {
    double value;
    if (source->valid) {
        if (sensors_get_value(source->chip, source->feature, &value)) {
            return 0;
        }
        if (t_type == FAHRENHEIT) {
            value = TO_FAHRENHEIT(value);
        } else if (t_type == KELVIN) {
            value = TO_KELVIN(value);
        }
        return (unsigned int)(value);
    } else {
        return 0;
    }
}

void temp_getusage(unsigned int *cpu_temp, unsigned int *sys_temp) {
    *cpu_temp = temp_read(&cpu_feature);
    *sys_temp = temp_read(&sys_feature);
}


dockapp.c
/*
 * Copyright (c) 1999 Alfredo K. Kojima
 * Copyright (c) 2001, 2002 Seiichi SATO <ssato@sh.rim.or.jp>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
 * AUTHOR BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
 * AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

 * This code is based on libdockapp-0.4.0
 * modified by Seiichi SATO <ssato@sh.rim.or.jp>
 */

#include <stdlib.h>
#include "dockapp.h"

#define WINDOWED_SIZE_W 64
#define WINDOWED_SIZE_H 64

/* global */
Display	*display = NULL;
Bool	dockapp_iswindowed = False;
Bool	dockapp_isbrokenwm = False;

/* private */
static Window	window = None;
static Window	icon_window = None;
static GC	gc = NULL;
static int	depth = 0;
static Atom	delete_win;
static int	width, height;
static int	offset_w, offset_h;

void
dockapp_open_window(char *display_specified, char *appname,
		    unsigned w, unsigned h, int argc, char **argv)
{
    XClassHint	    *classhint;
    XWMHints	    *wmhints;
    XTextProperty   title;
    XSizeHints	    sizehints;
    Window	    root;
    int		    ww, wh;

    /* Open Connection to X Server */
    display = XOpenDisplay(display_specified);
    if (!display) {
	fprintf(stderr, "%s: could not open display %s!\n", argv[0],
		XDisplayName(display_specified));
	exit(1);
    }
    root = DefaultRootWindow(display);

    width = w;
    height = h;

    if (dockapp_iswindowed) {
	offset_w = (WINDOWED_SIZE_W - w) / 2;
	offset_h = (WINDOWED_SIZE_H - h) / 2;
	ww = WINDOWED_SIZE_W;
	wh = WINDOWED_SIZE_H;
    } else {
	offset_w = offset_h = 0;
	ww = w;
	wh = h;
    }

    /* Create Windows */
    icon_window = XCreateSimpleWindow(display, root, 0, 0, ww, wh, 0, 0, 0);
    if (dockapp_isbrokenwm) {
	window = XCreateSimpleWindow(display, root, 0, 0, ww, wh, 0, 0, 0);
    } else {
	window = XCreateSimpleWindow(display, root, 0, 0, 1, 1, 0, 0, 0);
    }

    /* Set ClassHint */
    classhint = XAllocClassHint();
    if (classhint == NULL) {
	fprintf(stderr, "%s: can't allocate memory for wm hints!\n", argv[0]);
	exit(1);
    }
    classhint->res_class = "DockApp";
    classhint->res_name = appname;
    XSetClassHint(display, window, classhint);
    XFree(classhint);

    /* Set WMHints */
    wmhints = XAllocWMHints();
    if (wmhints == NULL) {
	fprintf(stderr, "%s: can't allocate memory for wm hints!\n", argv[0]);
	exit(1);
    }
    wmhints->flags = IconWindowHint | WindowGroupHint;
    if (!dockapp_iswindowed) {
	wmhints->flags |= StateHint;
	wmhints->initial_state = WithdrawnState;
    }
    wmhints->window_group = window;
    wmhints->icon_window = icon_window;
    XSetWMHints(display, window, wmhints);
    XFree(wmhints);

    /* Set WM Protocols */
    delete_win = XInternAtom(display, "WM_DELETE_WINDOW", False);
    XSetWMProtocols (display, icon_window, &delete_win, 1);

    /* Set Size Hints */
    sizehints.flags = USSize;
    if (!dockapp_iswindowed) {
	sizehints.flags |= USPosition;
	sizehints.x = sizehints.y = 0;
    } else {
	sizehints.flags |= PMinSize | PMaxSize;
	sizehints.min_width = sizehints.max_width = WINDOWED_SIZE_W;
	sizehints.min_height = sizehints.max_height = WINDOWED_SIZE_H;
    }
    sizehints.width = ww;
    sizehints.height = wh;
    XSetWMNormalHints(display, icon_window, &sizehints);

    /* Set WindowTitle for AfterStep Wharf */
    XStringListToTextProperty(&appname, 1, &title);
    XSetWMName(display, window, &title);
    XSetWMName(display, icon_window, &title);

    /* Set Command to start the app so it can be docked properly */
    XSetCommand(display, window, argv, argc);

    depth = DefaultDepth(display, DefaultScreen(display));
    gc = DefaultGC(display, DefaultScreen(display));

    XFlush(display);
}


void
dockapp_set_eventmask(long mask)
{
    XSelectInput(display, icon_window, mask);
    XSelectInput(display, window, mask);
}


static Pixmap
create_bg_pixmap(void)
{
    Pixmap bg;

    bg = XCreatePixmap(display, icon_window, WINDOWED_SIZE_W, WINDOWED_SIZE_H,
		       depth);
    XSetForeground(display, gc, dockapp_getcolor("rgb:ae/aa/ae"));
    XFillRectangle(display, bg, gc, 0, 0, WINDOWED_SIZE_W, WINDOWED_SIZE_H);
    XSetForeground(display, gc, dockapp_getcolor("rgb:ff/ff/ff"));
    XDrawLine(display, bg, gc, 0, 0, 0, 63);
    XDrawLine(display, bg, gc, 1, 0, 1, 62);
    XDrawLine(display, bg, gc, 2, 0, 63, 0);
    XDrawLine(display, bg, gc, 2, 1, 62, 1);
    XSetForeground(display, gc, dockapp_getcolor("rgb:52/55/52"));
    XDrawLine(display, bg, gc, 1, 63, 63, 63);
    XDrawLine(display, bg, gc, 2, 62, 63, 62);
    XDrawLine(display, bg, gc, 63, 1, 63, 61);
    XDrawLine(display, bg, gc, 62, 2, 62, 61);

    return bg;
}


void
dockapp_set_background(Pixmap pixmap)
{
    if (dockapp_iswindowed) {
	Pixmap bg;
	bg = create_bg_pixmap();
	XCopyArea(display, pixmap, bg, gc, 0, 0, width, height,
		  offset_w, offset_w);
	XSetWindowBackgroundPixmap(display, icon_window, bg);
	XSetWindowBackgroundPixmap(display, window, bg);
	XFreePixmap(display, bg);
    } else {
	XSetWindowBackgroundPixmap(display, icon_window, pixmap);
	XSetWindowBackgroundPixmap(display, window, pixmap);
    }
    XClearWindow(display, icon_window);
    XFlush(display);
}


void
dockapp_show(void)
{
    if (!dockapp_iswindowed)
	XMapRaised(display, window);
    else
	XMapRaised(display, icon_window);

    XFlush(display);
}


Bool
dockapp_xpm2pixmap(char **data, Pixmap *pixmap, Pixmap *mask,
		   XpmColorSymbol * colorSymbol, unsigned int nsymbols)
{
    XpmAttributes xpmAttr;
    xpmAttr.valuemask = XpmCloseness;
    xpmAttr.closeness = 40000;

    if (nsymbols) {
	xpmAttr.colorsymbols = colorSymbol;
	xpmAttr.numsymbols = nsymbols;
	xpmAttr.valuemask |= XpmColorSymbols;
    }

    if (XpmCreatePixmapFromData(display, icon_window, data, pixmap, mask, &xpmAttr) != 0)
	return False;

    return True;
}


Pixmap
dockapp_XCreatePixmap(int w, int h)
{
    return (XCreatePixmap(display, icon_window, w, h, depth));
}


void
dockapp_setshape(Pixmap mask, int x_ofs, int y_ofs)
{
    XShapeCombineMask(display, icon_window, ShapeBounding, -x_ofs, -y_ofs,
		      mask, ShapeSet);
    XShapeCombineMask(display, window, ShapeBounding, -x_ofs, -y_ofs,
		      mask, ShapeSet);
    XFlush(display);
}


void
dockapp_copyarea(Pixmap src, Pixmap dist, int x_src, int y_src, int w, int h,
		 int x_dist, int y_dist)
{
    XCopyArea(display, src, dist, gc, x_src, y_src, w, h, x_dist, y_dist);
}


void
dockapp_copy2window (Pixmap src)
{
    if (dockapp_isbrokenwm) {
	XCopyArea(display, src, window, gc, 0, 0, width, height, offset_w,
		  offset_h);
    } else {
	XCopyArea(display, src, icon_window, gc, 0, 0, width, height, offset_w,
		  offset_h);
    }
}


Bool
dockapp_nextevent_or_timeout(XEvent *event, unsigned long miliseconds)
{
    struct timeval timeout;
    fd_set rset;

    XSync(display, False);
    if (XPending(display)) {
	XNextEvent(display, event);
	return True;
    }

    timeout.tv_sec = miliseconds / 1000;
    timeout.tv_usec = (miliseconds % 1000) * 1000;

    FD_ZERO(&rset);
    FD_SET(ConnectionNumber(display), &rset);
    if (select(ConnectionNumber(display)+1, &rset, NULL, NULL, &timeout) > 0) {
	XNextEvent(display, event);
	if (event->type == ClientMessage) {
	    if (event->xclient.data.l[0] == delete_win) {
		XDestroyWindow(display,event->xclient.window);
		XCloseDisplay(display);
		exit(0);
	    }
	}
	if (dockapp_iswindowed) {
		event->xbutton.x -= offset_w;
		event->xbutton.y -= offset_h;
	}
	return True;
    }

    return False;
}


unsigned long
dockapp_getcolor(char *color_name)
{
    XColor color;

    if (!XParseColor(display, DefaultColormap(display, DefaultScreen(display)),
		     color_name, &color))
	fprintf(stderr, "can't parse color %s\n", color_name), exit(1);

    if (!XAllocColor(display, DefaultColormap(display, DefaultScreen(display)),
		     &color)) {
	fprintf(stderr, "can't allocate color %s. Using black\n", color_name);
	return BlackPixel(display, DefaultScreen(display));
    }

    return color.pixel;
}


unsigned long
dockapp_blendedcolor(char *color_name, int r, int g, int b, float fac)
{
    XColor color;

    if ((r < -255 || r > 255)||(g < -255 || g > 255)||(b < -255 || b > 255)){
	fprintf(stderr, "r:%d,g:%d,b:%d (r,g,b must be 0 to 255)", r, g, b);
	exit(1);
    }

    r *= 255;
    g *= 255;
    b *= 255;

    if (!XParseColor(display, DefaultColormap(display, DefaultScreen(display)),
		     color_name, &color))
	fprintf(stderr, "can't parse color %s\n", color_name), exit(1);

    if (!XAllocColor(display, DefaultColormap(display, DefaultScreen(display)),
		     &color)) {
	fprintf(stderr, "can't allocate color %s. Using black\n", color_name);
	return BlackPixel(display, DefaultScreen(display));
    }

    if (DefaultDepth(display, DefaultScreen(display)) < 16)
	return color.pixel;

    /* red */
    if (color.red + r > 0xffff) {
	color.red = 0xffff;
    } else if (color.red + r < 0) {
	color.red = 0;
    } else {
	color.red = (unsigned short)(fac * color.red + r);
    }

    /* green */
    if (color.green + g > 0xffff) {
	color.green = 0xffff;
    } else if (color.green + g < 0) {
	color.green = 0;
    } else {
	color.green = (unsigned short)(fac * color.green + g);
    }

    /* blue */
    if (color.blue + b > 0xffff) {
	color.blue = 0xffff;
    } else if (color.blue + b < 0) {
	color.blue = 0;
    } else {
	color.blue = (unsigned short)(fac * color.blue + b);
    }

    color.flags = DoRed | DoGreen | DoBlue;

    if (!XAllocColor(display, DefaultColormap(display, DefaultScreen(display)),
		     &color)) {
	fprintf(stderr, "can't allocate color %s. Using black\n", color_name);
	return BlackPixel(display, DefaultScreen(display));
    }

    return color.pixel;
}


main.c
/*    WMtemp - A dockapp to monitor CPU and MB temperatures
 *    Copyright (C) 2004  Peter Gnodde <peter@gnodde.org>
 *
 *    Some ideas loaned from the wmgtemp dockapp
 *
 *    Based on work by Mark Staggs <me@markstaggs.net>
 *    Copyright (C) 2002  Mark Staggs <me@markstaggs.net>
 *
 *    Based on work by Seiichi SATO <ssato@sh.rim.or.jp>
 *    Copyright (C) 2001,2002  Seiichi SATO <ssato@sh.rim.or.jp>

 *    This program is free software; you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation; either version 2 of the License, or
 *    (at your option) any later version.

 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.

 *    You should have received a copy of the GNU General Public License
 *    along with this program; if not, write to the Free Software
 *    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include <stdlib.h>
#include <string.h>

#include "dockapp.h"
#include "temp.h"
#include "xpm/xpm.h"

#define SIZE	    58
#define WINDOWED_BG "  \tc #AEAAAE"
#define MAX_HISTORY 16
#define CPUNUM_NONE -1

typedef enum { LIGHTON, LIGHTOFF } light;

Pixmap pixmap;
Pixmap backdrop_on;
Pixmap backdrop_off;
Pixmap parts;
Pixmap mask;
static char	*display_name = "";
static char	*light_color = NULL;	/* back-light color */
static unsigned update_interval = 1;
static light    backlight = LIGHTOFF;

static unsigned int cpu_temp = 0;
static unsigned int sys_temp = 0;
static unsigned int alarm_cpu = 60;
static unsigned int alarm_sys = 60;
char *configfile = NULL;

static char **backlight_on_xpm;
static char **backlight_off_xpm;

/* prototypes */
static void update(void);
static void switch_light(void);
static void draw_cpudigit(int per);
static void draw_sysdigit(int per);
static void parse_arguments(int argc, char **argv);
static void parse_sensor_path(sensor_path_t *path, char *argument);
static void print_help(char *prog);

int main(int argc, char **argv)
{
   XEvent event;
   XpmColorSymbol colors[2] = { {"Back0", NULL, 0}, {"Back1", NULL, 0} };
   int ncolor = 0;

   /* Parse CommandLine */
   parse_arguments(argc, argv);

   if (t_type == FAHRENHEIT) {
	   alarm_cpu = TO_FAHRENHEIT(alarm_cpu);
	   alarm_sys = TO_FAHRENHEIT(alarm_sys);

	   backlight_on_xpm = fahrenheit_on_xpm;
	   backlight_off_xpm = fahrenheit_off_xpm;
   } else if (t_type == KELVIN) {
	   alarm_cpu = TO_KELVIN(alarm_cpu);
	   alarm_sys = TO_KELVIN(alarm_sys);

	   backlight_on_xpm = kelvin_on_xpm;
	   backlight_off_xpm = kelvin_off_xpm;
   } else {
	   backlight_on_xpm = celcius_on_xpm;
	   backlight_off_xpm = celcius_off_xpm;
   }
   
   /* Initialize Application */
   temp_init(configfile);
   dockapp_open_window(display_name, PACKAGE, SIZE, SIZE, argc, argv);
   dockapp_set_eventmask(ButtonPressMask);

   if(light_color)
   {
      colors[0].pixel = dockapp_getcolor(light_color);
      colors[1].pixel = dockapp_blendedcolor(light_color, -24, -24, -24, 1.0);
      ncolor = 2;
   }

   /* change raw xpm data to pixmap */
   if(dockapp_iswindowed)
      backlight_on_xpm[1] = backlight_off_xpm[1] = WINDOWED_BG;

   if(!dockapp_xpm2pixmap(backlight_on_xpm, &backdrop_on, &mask, colors, ncolor))
   {
      fprintf(stderr, "Error initializing backlit background image.\n");
      exit(1);
   }
   if(!dockapp_xpm2pixmap(backlight_off_xpm, &backdrop_off, NULL, NULL, 0))
   {
      fprintf(stderr, "Error initializing background image.\n");
      exit(1);
   }
   if(!dockapp_xpm2pixmap(parts_xpm, &parts, NULL, colors, ncolor))
   {
      fprintf(stderr, "Error initializing parts image.\n");
      exit(1);
   }

   /* shape window */
   if(!dockapp_iswindowed)
      dockapp_setshape(mask, 0, 0);
   if(mask) XFreePixmap(display, mask);

   /* pixmap : draw area */
   pixmap = dockapp_XCreatePixmap(SIZE, SIZE);

   /* Initialize pixmap */
   if(backlight == LIGHTON) 
      dockapp_copyarea(backdrop_on, pixmap, 0, 0, SIZE, SIZE, 0, 0);
   else
      dockapp_copyarea(backdrop_off, pixmap, 0, 0, SIZE, SIZE, 0, 0);

   dockapp_set_background(pixmap);
   dockapp_show();

   update();

   /* Main loop */
   while(1)
   {
      if (dockapp_nextevent_or_timeout(&event, update_interval * 1000))
      {
         /* Next Event */
         switch(event.type)
         {
            case ButtonPress:
               switch_light();
               break;
            default: /* make gcc happy */
               break;
         }
      }
      else
      {
         /* Time Out */
         update();
      }
   }

   return 0;
}

/* called by timer */
static void update(void)
{
   static light pre_backlight;
   static Bool in_alarm_mode = False;

   /* get current cpu usage in percent */
   temp_getusage(&cpu_temp, &sys_temp);

   /* alarm mode */
   if(cpu_temp >= alarm_cpu || sys_temp >= alarm_sys) 
   {
      if(!in_alarm_mode)
      {
         in_alarm_mode = True;
         pre_backlight = backlight;
      }
      if(backlight == LIGHTOFF)
      {
         switch_light();
         return;
      }
   } 
   else
   {
      if(in_alarm_mode)
      {
         in_alarm_mode = False;
         if (backlight != pre_backlight) 
         {
            switch_light();
            return;
         }
      }
   }

   /* all clear */
   if (backlight == LIGHTON) 
      dockapp_copyarea(backdrop_on, pixmap, 0, 0, 58, 58, 0, 0);
   else 
      dockapp_copyarea(backdrop_off, pixmap, 0, 0, 58, 58, 0, 0);

   /* draw digit */
   draw_cpudigit(cpu_temp);
   draw_sysdigit(sys_temp);

   /* show */
   dockapp_copy2window(pixmap);
}

/* called when mouse button pressed */
static void switch_light(void)
{
   switch (backlight)
   {
      case LIGHTOFF:
         backlight = LIGHTON;
         dockapp_copyarea(backdrop_on, pixmap, 0, 0, 58, 58, 0, 0);
         break;
      case LIGHTON:
         backlight = LIGHTOFF;
         dockapp_copyarea(backdrop_off, pixmap, 0, 0, 58, 58, 0, 0);
         break;
   }

   /* redraw digit */
   temp_getusage(&cpu_temp, &sys_temp);
   draw_cpudigit(cpu_temp);
   draw_sysdigit(sys_temp);

   /* show */
   dockapp_copy2window(pixmap);
}

static void draw_cpudigit(int per)
{
   int v100, v10, v1;
   int y = 0;

   if (per < 0) per = 0;

   v100 = per / 100;
   v10  = (per - v100 * 100) / 10;
   v1   = (per - v100 * 100 - v10 * 10);

   if (backlight == LIGHTON) y = 20;

   /* draw digit */
   dockapp_copyarea(parts, pixmap, v1 * 10, y, 10, 20, 29, 7);
   if (v10 != 0 || v100 != 0) {
	   dockapp_copyarea(parts, pixmap, v10 * 10, y, 10, 20, 17, 7);
	   
	   if (v100 != 0) {
		   dockapp_copyarea(parts, pixmap, v100 * 10, y, 10, 20,  5, 7);
	   }
   }
}


static void draw_sysdigit(int per)
{
   int v100, v10, v1;
   int y = 0;

   if (per < 0) per = 0;

   v100 = per / 100;
   v10  = (per - v100 * 100) / 10;
   v1   = (per - v100 * 100 - v10 * 10);

   if (backlight == LIGHTON) y = 20;

   /* draw digit */
   dockapp_copyarea(parts, pixmap, v1 * 10, y, 10, 20, 29, 34);
   if (v10 != 0 || v100 != 0) {
	   dockapp_copyarea(parts, pixmap, v10 * 10, y, 10, 20, 17, 34);

	   if (v100 != 0) {
		   dockapp_copyarea(parts, pixmap, v100 * 10, y, 10, 20, 5, 34);
	   }
   }
}

static void parse_arguments(int argc, char **argv)
{
   int i;
   int integer;
   for (i = 1; i < argc; i++) 
   {
      if (!strcmp(argv[i], "-v"))
         printf("%s version %s\n", PACKAGE, VERSION), exit(0);
      else if (!strcmp(argv[i], "-d")) 
      {
         display_name = argv[i + 1];
         i++;
      }
      else if (!strcmp(argv[i], "-ac")) 
      {
         if (argc == i + 1)
            alarm_cpu = 60;
         else if (sscanf(argv[i + 1], "%i", &integer) != 1)
            alarm_cpu = 60;
         else if (integer < 0 || integer > 100)
            fprintf(stderr, "%s: argument %s must be from 0 to 100\n",
                    argv[0], argv[i]), exit(1);
         else
            alarm_cpu = integer, i++;
      }
      else if (!strcmp(argv[i], "-as")) 
      {
         if (argc == i + 1)
            alarm_sys = 60;
         else if (sscanf(argv[i + 1], "%i", &integer) != 1)
            alarm_sys = 60;
         else if (integer < 0 || integer > 100)
            fprintf(stderr, "%s: argument %s must be from 0 to 100\n",
                    argv[0], argv[i]), exit(1);
         else
            alarm_sys = integer, i++;
      }
      else if (!strcmp(argv[i], "-bl"))
         backlight = LIGHTON;
      else if (!strcmp(argv[i], "-lc")) 
      {
         light_color = argv[i + 1];
         i++;
      }
      else if (!strcmp(argv[i], "-i")) 
      {
         if (argc == i + 1)
            fprintf(stderr, "%s: error parsing argument for option %s\n",
                    argv[0], argv[i]), exit(1);
         if (sscanf(argv[i + 1], "%i", &integer) != 1)
             fprintf(stderr, "%s: error parsing argument for option %s\n",
                     argv[0], argv[i]), exit(1);
         if (integer < 1)
             fprintf(stderr, "%s: argument %s must be >=1\n",
                     argv[0], argv[i]), exit(1);
         update_interval = integer;
         i++;
      }
      else if (!strcmp(argv[i], "-w"))
         dockapp_iswindowed = True;
      else if (!strcmp(argv[i], "-bw"))
         dockapp_isbrokenwm = True;
	  else if (!strcmp(argv[i], "-c")) {
		  if (argc == i + 1) {
			fprintf(stderr, "%s: error parsing argument for option %s\n", argv[0], argv[i]);
			exit(1);
		  }

		  configfile = argv[i + 1];
		  i++;
	  }
	  else if (!strcmp(argv[i], "-s")) {
		  sensor_path_t t;

		  t = cpu_sensor_path;
		  cpu_sensor_path = sys_sensor_path;
		  sys_sensor_path = t;
	  }
	  else if (!strcmp(argv[i], "-cf")) {
		  if (argc == i + 1) {
			fprintf(stderr, "%s: error parsing argument for option %s\n", argv[0], argv[i]);
			exit(1);
		  }

		  cpu_sensor_path.subfeature = argv[i + 1];
		  i++;
	  }
	  else if (!strcmp(argv[i], "-sf")) {
		  if (argc == i + 1) {
			fprintf(stderr, "%s: error parsing argument for option %s\n", argv[0], argv[i]);
			exit(1);
		  }

		  sys_sensor_path.subfeature = argv[i + 1];
		  i++;
	  }
	  else if (!strcmp(argv[i], "-cp") || !strcmp(argv[i], "-sp")) {
		  if (argc == i + 1) {
			fprintf(stderr, "%s: error parsing argument for option %s\n", argv[0], argv[i]);
			exit(1);
		  }
          parse_sensor_path(argv[i][1] == 'c' ? &cpu_sensor_path : &sys_sensor_path,
				  argv[i + 1]);
		  i++;
	  }
	  else if (!strcmp(argv[i], "-f"))
		  t_type = FAHRENHEIT;
	  else if (!strcmp(argv[i], "-k"))
		  t_type = KELVIN;
      else
      {
         print_help(argv[0]), exit(1);
      }
   }
}

static void parse_sensor_path(sensor_path_t *path, char *argument)
{
	char const ** parts[] = { &path->chip, &path->feature, &path->subfeature };
	int i = 0;
	char * next;
	while (i < 3) {
		next = strchr(argument, '/');
		if (next != argument) {
			*parts[i] = argument;
		}
		if (!next) {
			break;
		}
		*next = '\0';
		argument = next + 1;
		i ++;
	}
}

static void print_help(char *prog)
{
   printf("Usage : %s [OPTIONS]\n", prog);
   printf("WMtemp - temperature monitor dockapp\n");
   printf("  -d <string>                    display to use\n");
   printf("  -bl                            turn on back-light\n");
   printf("  -lc <string>                   back-light color (rgb:6E/C6/3B is default)\n");
   printf("  -i <number>                    number of secs between updates (1 is default)\n");
   printf("  -h                             show this help text and exit\n");
   printf("  -v                             show program version and exit\n");
   printf("  -w                             run the application in windowed mode\n");
   printf("  -bw                            activate broken window manager fix\n");
   printf("  -ac <number>                   activate alarm mode of CPU\n");
   printf("                                 <number> is threshold from 0 to 100\n");
   printf("                                 (60 is default)\n");
   printf("  -as <number>                   activate alarm mode of system\n");
   printf("                                 <number> is threshold from 0 to 100\n");
   printf("                                 (60 is default)\n");
   printf("  -c <path>                      location of the sensors.conf file\n");
   printf("                                 ('/etc/sensors.conf' is default)\n");
   printf("  -cf <feature>                  which feature to use for cpu temperature\n");
   printf("                                 ('temp1_input' is default)\n");
   printf("  -sf <feature>                  which feature to use for sys temperature\n");
   printf("                                 ('temp2_input' is default)\n");
   printf("  -cp <chip/feature/subft.>      full cpu temperature sensor path\n");
   printf("                                 ('//temp1_input' is default)\n");
   printf("  -sp <chip/feature/subft.>      full system temperature sensor path\n");
   printf("                                 ('//temp2_input' is default)\n");
   printf("  -s                             swap the cpu and sys temperatures\n");
   printf("                                 (/etc/sensors.conf is default)\n");
   printf("  -f                             show temperatures in Fahrenheit\n");
   printf("  -k                             show temperatures in Kelvin\n");
}
